Date: Mon, 24 Apr 2017 17:22:02 +0000

From: 78559bc5db3495f12b0284e4be5fc8bfcd7041e2 Mon Sep 17 00:00:00 2001

Subject: [PATCH] XRE-11686 : gstreamer error seen with super8 encrypted streams

	qtdemux: also push buffers without encryption info instead of dropping them
	added senc parser in qtdemux fixed OpenCDMdecryptor in WPE to not call decrypt() when totalEncryptedBytes=0

Source: GOOGLE


Upstream-Status: Submitted [https://coral.googlesource.com/imx-gst-plugins-good/+/762e9c645ec13513c62eb5a3800d7406e01cdcb7 ]

Signed-off-by: Enrique Ocaña González <eocanha@igalia.com>
---


 gst/isomp4/qtdemux.c | 26 ++++++++++++--------------
 1 file changed, 12 insertions(+), 14 deletions(-)

diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index 6b0c820..7b71237 100644
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -5421,20 +5421,18 @@ gst_qtdemux_decorate_and_push_buffer (GstQTDemux * qtdemux,
         gst_pad_push_event (stream->pad, event);
     }
 
-    if (info->crypto_info == NULL) {
-      GST_DEBUG_OBJECT (qtdemux, "cenc metadata hasn't been parsed yet");
-      gst_buffer_unref (buf);
-      goto exit;
-    }
-
-    index = stream->sample_index - (stream->n_samples - info->crypto_info->len);
-    if (G_LIKELY (index >= 0 && index < info->crypto_info->len)) {
-      /* steal structure from array */
-      crypto_info = g_ptr_array_index (info->crypto_info, index);
-      g_ptr_array_index (info->crypto_info, index) = NULL;
-      GST_LOG_OBJECT (qtdemux, "attaching cenc metadata [%u]", index);
-      if (!crypto_info || !gst_buffer_add_protection_meta (buf, crypto_info))
-        GST_ERROR_OBJECT (qtdemux, "failed to attach cenc metadata to buffer");
+    if (info->crypto_info == NULL)
+      GST_DEBUG_OBJECT (qtdemux, "cenc metadata hasn't been parsed yet, pushing buffer as if it wasn't encrypted");
+    else {
+      index = stream->sample_index - (stream->n_samples - info->crypto_info->len);
+      if (G_LIKELY (index >= 0 && index < info->crypto_info->len)) {
+        /* steal structure from array */
+        crypto_info = g_ptr_array_index (info->crypto_info, index);
+        g_ptr_array_index (info->crypto_info, index) = NULL;
+        GST_LOG_OBJECT (qtdemux, "attaching cenc metadata [%u]", index);
+        if (!crypto_info || !gst_buffer_add_protection_meta (buf, crypto_info))
+          GST_ERROR_OBJECT (qtdemux, "failed to attach cenc metadata to buffer");
+      }
     }
   }
 
-- 
1.8.3.2

